### PR TITLE
Use babel's lazy import mode for faster startup

### PR DESCRIPTION
**Pull request by devongovett**
This uses the [lazy](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs#lazy) option of `@babel/plugin-transform-modules-commonjs` to lazy load imports rather than eagerly load them. This dramatically reduces the initial startup time of Parcel by avoiding node's `require` cost until necessary.

On the kitchen sink demo, this reduced a 100% cached build from ~1.2s to ~260ms. 🥳

### CODE REVIEW

File: packages/dev/babel-preset/index.js [hunk 1]
```diff
@@ -15,6 +16,12 @@ module.exports = () => ({
     require('@babel/plugin-proposal-class-properties'),
     require('@babel/plugin-proposal-nullish-coalescing-operator'),
     require('@babel/plugin-proposal-optional-chaining'),
+    [
+      require('@babel/plugin-transform-modules-commonjs'),
+      {
+        lazy: () => process.env.PARCEL_BUILD_ENV !== 'test',
```
Comment (devongovett) [inline packages/dev/babel-preset/index.js:22]: Not entirely sure why, but lazy mode seems to hang the tests completely. Maybe something with mocha?

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 8.02s | +878.00ms ⚠️ |
| Cached | 1.50s | -937.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 228.00ms | -52.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 229.00ms | -52.00ms 🚀 |
| dist/legacy/index.45cccf70.js | 1.10kb | +0.00b | 2.36s | +681.00ms ⚠️ |
| dist/modern/index.1964d1f6.js | 1.10kb | +0.00b | 2.36s | +1.76s ⚠️ |
| dist/legacy/index.html | 701.00b | +0.00b | 1.95s | +123.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 1.94s | +117.00ms ⚠️ |
| dist/legacy/index.b3acc8eb.css | 77.00b | +0.00b | 1.76s | +195.00ms ⚠️ |
| dist/modern/index.7bc22a8b.css | 77.00b | +0.00b | 2.35s | +792.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 56.00ms | -131.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 56.00ms | -131.00ms 🚀 |
| dist/legacy/index.45cccf70.js | 1.10kb | +0.00b | 122.00ms | -245.00ms 🚀 |
| dist/modern/index.1964d1f6.js | 1.10kb | +0.00b | 121.00ms | -245.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 142.00ms | -227.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 121.00ms | -245.00ms 🚀 |
| dist/legacy/index.b3acc8eb.css | 77.00b | +0.00b | 122.00ms | -245.00ms 🚀 |
| dist/modern/index.7bc22a8b.css | 77.00b | +0.00b | 121.00ms | -245.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 32.94s | +1000.00ms |
| Cached | 1.71s | -862.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 1.20s | +1.03s ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 478.68kb | +0.00b | 72.00ms | -249.00ms 🚀 |
| dist/PermalinkedComment.9c5ad065.js | 4.22kb | +0.00b | 62.00ms | -244.00ms 🚀 |
| dist/UserProfile.c1927888.js | 1.62kb | +0.00b | 62.00ms | -244.00ms 🚀 |
| dist/NotFound.f8a0616d.js | 387.00b | +0.00b | 62.00ms | -244.00ms 🚀 |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 12.00ms | -181.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.27m | +1.74s |
| Cached | 4.61s | -924.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.474ded24.js | 2.72mb | +4.00b ⚠️ | 1.73m | +1.78s |
| dist/media-viewer.bcba38b3.js | 71.96kb | +0.00b | 20.83s | -1.24s 🚀 |
| dist/Modal.19ffed32.js | 41.27kb | +0.00b | 7.45s | -790.00ms 🚀 |
| dist/component.d7c29980.js | 30.70kb | +0.00b | 9.16s | +6.05s ⚠️ |
| dist/component.54080a26.js | 22.50kb | +0.00b | 7.45s | -790.00ms 🚀 |
| dist/js.5733a017.js | 16.40kb | +0.00b | 7.45s | -793.00ms 🚀 |
| dist/workerHasher.1e706e77.js | 11.75kb | +0.00b | 17.26s | +2.04s ⚠️ |
| dist/EmojiTypeAheadComponent.1990d771.js | 8.67kb | +0.00b | 58.56s | +43.34s ⚠️ |
| dist/component.77c0f682.js | 6.05kb | +0.00b | 9.91s | +1.67s ⚠️ |
| dist/card.383bd9bd.js | 5.61kb | +0.00b | 20.83s | +3.02s ⚠️ |
| dist/media-viewer.59c17dad.js | 3.92kb | +0.00b | 20.83s | -1.18s 🚀 |
| dist/png-chunks-extract.a844fdb6.js | 3.41kb | +0.00b | 7.45s | -791.00ms 🚀 |
| dist/dropzone.3c67f0a6.js | 3.23kb | +0.00b | 38.21s | -22.30s 🚀 |
| dist/Modal.0304b9ed.js | 3.08kb | +0.00b | 7.45s | +1.63s ⚠️ |
| dist/16.290b4a43.js | 2.38kb | +0.00b | 2.21s | -904.00ms 🚀 |
| dist/EmojiUploadComponent.d54a1fc6.js | 2.18kb | +0.00b | 58.56s | +43.34s ⚠️ |
| dist/card.eea6ad8c.js | 2.03kb | +0.00b | 38.21s | +16.14s ⚠️ |
| dist/ResourcedEmojiComponent.abe8596c.js | 2.00kb | +0.00b | 9.06s | +3.61s ⚠️ |
| dist/16.d16f14ed.js | 1.73kb | +0.00b | 8.19s | -535.00ms 🚀 |
| dist/16.1b6ca79f.js | 1.66kb | +0.00b | 7.89s | -1.24s 🚀 |
| dist/browser.75813db0.js | 1.61kb | +0.00b | 1.08m | -37.34s 🚀 |
| dist/workerHasher.6d5b22d7.js | 1.54kb | +0.00b | 58.56s | +43.34s ⚠️ |
| dist/workerHasher.fa198233.js | 1.54kb | +0.00b | 1.73m | +39.92s ⚠️ |
| dist/16.19c93285.js | 1.33kb | +0.00b | 7.89s | -1.24s 🚀 |
| dist/16.c7751e34.js | 1.33kb | +0.00b | 7.89s | -561.00ms 🚀 |
| dist/16.51049e2c.js | 1.32kb | +0.00b | 2.19s | -917.00ms 🚀 |
| dist/16.436f097f.js | 1.21kb | +0.00b | 2.21s | -899.00ms 🚀 |
| dist/16.b46ef690.js | 1.14kb | +0.00b | 2.19s | -917.00ms 🚀 |
| dist/16.9ba230ea.js | 1.13kb | +0.00b | 7.89s | -1.51s 🚀 |
| dist/16.eafe1a11.js | 1.13kb | +0.00b | 7.89s | -1.24s 🚀 |
| dist/heading2.a2a1be3f.js | 1.11kb | +0.00b | 11.20s | -45.09s 🚀 |
| dist/16.3920766f.js | 1.10kb | +0.00b | 2.19s | -3.26s 🚀 |
| dist/component.6600f83e.js | 1.07kb | +0.00b | 9.06s | +3.61s ⚠️ |
| dist/panel-warning.97a0e3b3.js | 1008.00b | +0.00b | 11.06s | -44.91s 🚀 |
| dist/panel-success.9b923e58.js | 845.00b | +0.00b | 11.06s | -44.91s 🚀 |
| dist/panel-note.97a4d541.js | 841.00b | +0.00b | 9.07s | -46.90s 🚀 |
| dist/media-picker-analytics-error-boundary.f8ddfee3.js | 839.00b | +0.00b | 1.08m | +4.19s ⚠️ |
| dist/media-viewer-analytics-error-boundary.0afa7031.js | 755.00b | +0.00b | 20.83s | -1.17s 🚀 |
| dist/EmojiPickerComponent.2e54b9c6.js | 535.00b | +0.00b | 57.82s | +42.60s ⚠️ |
| dist/index.html | 119.00b | +0.00b | 5.59s | +4.07s ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.474ded24.js | 2.72mb | +0.00b | 161.00ms | -459.00ms 🚀 |
| dist/pdfRenderer.15a6a023.js | 1.11mb | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/editorView.e06544fb.js | 587.12kb | +0.00b | 155.00ms | -462.00ms 🚀 |
| dist/popup.6c453e3c.js | 169.96kb | +0.00b | 156.00ms | -460.00ms 🚀 |
| dist/Toolbar.83ac0b6e.js | 92.34kb | +0.00b | 155.00ms | -462.00ms 🚀 |
| dist/EmojiPickerComponent.ae601247.js | 75.29kb | +0.00b | 156.00ms | -462.00ms 🚀 |
| dist/media-viewer.bcba38b3.js | 71.96kb | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/card.645a5c17.js | 55.01kb | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/card.0313bac8.js | 52.78kb | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/Modal.19ffed32.js | 41.27kb | +0.00b | 160.00ms | -461.00ms 🚀 |
| dist/component.d7c29980.js | 30.70kb | +0.00b | 160.00ms | -461.00ms 🚀 |
| dist/component.54080a26.js | 22.50kb | +0.00b | 160.00ms | -460.00ms 🚀 |
| dist/DatePicker.0df05dac.js | 21.21kb | +0.00b | 156.00ms | -462.00ms 🚀 |
| dist/EmojiPickerComponent.7dc0ebd0.js | 18.23kb | +0.00b | 156.00ms | -462.00ms 🚀 |
| dist/dropzone.7c704bdb.js | 17.27kb | +0.00b | 156.00ms | -460.00ms 🚀 |
| dist/smartMediaEditor.9623570c.js | 16.87kb | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/js.5733a017.js | 16.40kb | +0.00b | 160.00ms | -460.00ms 🚀 |
| dist/ui.fee77548.js | 13.97kb | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/workerHasher.1e706e77.js | 11.75kb | +0.00b | 156.00ms | -462.00ms 🚀 |
| dist/EmojiTypeAheadComponent.1990d771.js | 8.67kb | +0.00b | 156.00ms | -462.00ms 🚀 |
| dist/component.77c0f682.js | 6.05kb | +0.00b | 160.00ms | -460.00ms 🚀 |
| dist/card.383bd9bd.js | 5.61kb | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/media-viewer.59c17dad.js | 3.92kb | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/EmojiPickerComponent.627aca6b.js | 3.58kb | +0.00b | 156.00ms | -462.00ms 🚀 |
| dist/index.ac562c80.css | 3.46kb | +0.00b | 155.00ms | -462.00ms 🚀 |
| dist/png-chunks-extract.a844fdb6.js | 3.41kb | +0.00b | 160.00ms | -460.00ms 🚀 |
| dist/dropzone.3c67f0a6.js | 3.23kb | +0.00b | 156.00ms | -460.00ms 🚀 |
| dist/Modal.0304b9ed.js | 3.08kb | +0.00b | 160.00ms | -461.00ms 🚀 |
| dist/clipboard.da112b6e.js | 2.75kb | +0.00b | 156.00ms | -460.00ms 🚀 |
| dist/16.290b4a43.js | 2.38kb | +0.00b | 160.00ms | -461.00ms 🚀 |
| dist/EmojiUploadComponent.d54a1fc6.js | 2.18kb | +0.00b | 156.00ms | -462.00ms 🚀 |
| dist/card.eea6ad8c.js | 2.03kb | +0.00b | 155.00ms | -461.00ms 🚀 |
| dist/ResourcedEmojiComponent.abe8596c.js | 2.00kb | +0.00b | 160.00ms | -461.00ms 🚀 |
| dist/date.231500d2.js | 1.73kb | +0.00b | 158.00ms | -461.00ms 🚀 |
| dist/16.d16f14ed.js | 1.73kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/images.80c92f5e.js | 1.67kb | +0.00b | 157.00ms | -463.00ms 🚀 |
| dist/16.1b6ca79f.js | 1.66kb | +0.00b | 159.00ms | -461.00ms 🚀 |
| dist/feedback.d5590b34.js | 1.64kb | +0.00b | 157.00ms | -461.00ms 🚀 |
| dist/browser.75813db0.js | 1.61kb | +0.00b | 156.00ms | -460.00ms 🚀 |
| dist/workerHasher.6d5b22d7.js | 1.54kb | +0.00b | 156.00ms | -462.00ms 🚀 |
| dist/workerHasher.fa198233.js | 1.54kb | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/list-number.c35f9589.js | 1.46kb | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/status.8ab4226b.js | 1.46kb | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/16.9ccb5b28.js | 1.38kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/code.3dd71280.js | 1.38kb | +0.00b | 158.00ms | -461.00ms 🚀 |
| dist/16.19c93285.js | 1.33kb | +0.00b | 160.00ms | -460.00ms 🚀 |
| dist/16.c7751e34.js | 1.33kb | +0.00b | 159.00ms | -461.00ms 🚀 |
| dist/16.51049e2c.js | 1.32kb | +0.00b | 160.00ms | -461.00ms 🚀 |
| dist/16.18344f47.js | 1.31kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/link.2f4427e6.js | 1.30kb | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/heading6.fdbbdf4d.js | 1.30kb | +0.00b | 157.00ms | -461.00ms 🚀 |
| dist/heading3.2113b7a1.js | 1.29kb | +0.00b | 157.00ms | -461.00ms 🚀 |
| dist/16.2c4c1846.js | 1.27kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/emoji.93ae61d3.js | 1.23kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/16.d17ab186.js | 1.23kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/16.436f097f.js | 1.21kb | +0.00b | 160.00ms | -461.00ms 🚀 |
| dist/16.b4da0acb.js | 1.19kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/16.02c68c4f.js | 1.18kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/heading5.e195aab1.js | 1.17kb | +0.00b | 157.00ms | -461.00ms 🚀 |
| dist/expand.450f3aea.js | 1.16kb | +0.00b | 156.00ms | -462.00ms 🚀 |
| dist/media-card-analytics-error-boundary.2323af57.js | 1.15kb | +0.00b | 156.00ms | -460.00ms 🚀 |
| dist/16.b46ef690.js | 1.14kb | +0.00b | 160.00ms | -461.00ms 🚀 |
| dist/16.4e3b0235.js | 1.14kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/16.9ba230ea.js | 1.13kb | +0.00b | 159.00ms | -461.00ms 🚀 |
| dist/16.a37df53b.js | 1.13kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/16.9adb4ec7.js | 1.13kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/16.eafe1a11.js | 1.13kb | +0.00b | 160.00ms | -460.00ms 🚀 |
| dist/heading2.a2a1be3f.js | 1.11kb | +0.00b | 157.00ms | -461.00ms 🚀 |
| dist/16.3920766f.js | 1.10kb | +0.00b | 160.00ms | -461.00ms 🚀 |
| dist/component.6600f83e.js | 1.07kb | +0.00b | 160.00ms | -461.00ms 🚀 |
| dist/mention.4d64e059.js | 1.07kb | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/heading4.ec980e93.js | 1.06kb | +0.00b | 157.00ms | -461.00ms 🚀 |
| dist/16.afcb09b6.js | 1.05kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/layout.52ba15f7.js | 1.05kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/divider.16defde6.js | 1.03kb | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/quote.aa25b83c.js | 1.02kb | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/action.9aa72b39.js | 1.00kb | +0.00b | 158.00ms | -462.00ms 🚀 |
| dist/decision.20244046.js | 1010.00b | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/panel-warning.97a0e3b3.js | 1008.00b | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/list.c9147a62.js | 978.00b | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/heading1.7d899ba5.js | 976.00b | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/panel-error.8d11e8b7.js | 906.00b | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/panel.a28c56ee.js | 897.00b | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/table.5a4874ee.js | 889.00b | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/panel-success.9b923e58.js | 845.00b | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/panel-note.97a4d541.js | 841.00b | +0.00b | 157.00ms | -462.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.f8ddfee3.js | 839.00b | +0.00b | 156.00ms | -460.00ms 🚀 |
| dist/media-card-analytics-error-boundary.663369b9.js | 835.00b | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/media-card-analytics-error-boundary.73ba4074.js | 835.00b | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.0afa7031.js | 755.00b | +0.00b | 155.00ms | -462.00ms 🚀 |
| dist/EmojiPickerComponent.2e54b9c6.js | 535.00b | +0.00b | 156.00ms | -462.00ms 🚀 |
| dist/simpleHasher.16c8bf41.js | 528.00b | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/simpleHasher.a29b2a8e.js | 528.00b | +0.00b | 156.00ms | -461.00ms 🚀 |
| dist/index.html | 119.00b | +0.00b | 194.00ms | -428.00ms 🚀 |

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/275959264216875528)**

**Comment by mischnic** [comment]
> On the kitchen sink demo, this reduced a 100% cached build from ~1.2s to ~260ms. 🥳

Is that with babel-register or the production build?

**Comment by mischnic** [comment]
Is this require even called now?  https://github.com/parcel-bundler/parcel/blob/4376130f8e8307187e857c167e26e2481814de2a/packages/core/parcel/src/cli.js#L12

**Comment by devongovett** [comment]
> Is that with babel-register or the production build?

production

> Is this require even called now?

yes? how do you mean?

File: packages/core/fs/src/index.js [hunk 2]
```diff
@@ -4,9 +4,9 @@ import type {FilePath} from '@parcel/types';
 import path from 'path';
 
 export type * from './types';
-export {NodeFS} from './NodeFS';
-export {MemoryFS} from './MemoryFS';
-export {OverlayFS} from './OverlayFS';
+export * from './NodeFS';
+export * from './MemoryFS';
+export * from './OverlayFS';
```
Comment (mischnic) [inline packages/core/fs/src/index.js:9]: In most index files, you removed the namespace export but you added it in this case?

**Comment by mischnic** [comment]
> yes? how do you mean?

I was under the impression that that plugin processed every `require` calls, but it's only ESM module declarations anyway...

File: packages/core/fs/src/index.js (hunk 2 above)
Comment (devongovett) [reply packages/core/fs/src/index.js:9]: Yeah... it's because we rely on these being executed immediately since they register types with the serializer. 😞

**Review by mischnic — APPROVED** [review]

**Pull request merged**